lib/handler/connect.c

```diff
@@ -56,6 +56,7 @@ struct st_connect_generator_t {
         struct st_server_address_t list[MAX_ADDRESSES_PER_FAMILY * 2];
         size_t size;
         size_t used;
+        const struct st_server_address_t *trying;
```

Comment:
Is this field ever going to be different from `server_addresses.list[server_addresses.used]`?

Reply:
Regarding how it works, I think you observation is correct.

Regarding if it is a behavior that we can depend on, I think it is, due to following reasons. Maybe what we need is a doc-comment around the structure regarding how the data structure is used:
* Of the entries in `st_connect_generator_t::server_addresses`, elements [0, used) contain addresses to which we have attempted to (or currently under an attempt to) connect.
* [used, size) contain addresses to which we have not.
* Considering that we need to at least have a list of addresses to which we might try to connect in the future, it makes sense to have this type of data structure.
* Due to us dividing an array into two logical lists, when trying a new address, it makes sense to pick the one at the border of the two lists.
* When trying to connect to a new address, we want to pick an address in the other address family if such an address exists. That's the logic around `pick_and_swap`.

FWIW, I might agree that It might seem fragile to rely on `used` to determine the address to which we have been currently trying to connect to. But if that is the concern, I might argue that the proposed approach (of retaining a pointer to the last address being used) is also fragile. Both of the approaches assume that the addresses that we have tried are not moved, and that only one address is being tried at a time. If we are to implement Happy Eyeballs in full spec, we would not have the second guarantee.

Therefore, if we want something solid, I think we should retain a copy of the address to which we are trying to connect to, rather than a pointer to an entity containing such address.

That said, I am not sure if we need such complexity when existing structure is capable of telling what the address is. Hence my original comment.

Reply:
An alternative implementation that removes the new member is in #3163.

`get_dest_addr_str` optionally returns the value so that the probe `connect_success` probe can return it.  If we do not want the detail in the probe (or we do not even want the `connect_success` probe, then `get_dest_addr_str` can be simplified and it will be an even smaller change.

Please let me know if think that that is a better trade-off to do it that way than to add `trying` to the state.

Reply:
Done in f47fd98b1.

`get_dest_addr_str` optionally returns the `struct st_server_address_t` pointer so that the probe `connect_success` can log it.  (If we do not want the new probe, then `get_dest_addr_str` can be simplified and the PR will be an even smaller change.)